src/video/wayland/SDL_waylandwindow.c

```diff
@@ -448,8 +491,102 @@ static const struct xdg_toplevel_listener toplevel_listener_xdg = {
+    SDL_Window *window = wind->sdlwindow;
+    int width, height;
+    enum libdecor_window_state window_state;
+    struct libdecor_state *state;
+
+    /* window size */
+    if (!libdecor_configuration_get_content_size(configuration, frame, &width, &height)) {
+        width = window->w;
+        height = window->h;
+    }
 
+    wind->resize.width = width;
+    wind->resize.height = height;
+
+    wind->resize.pending = SDL_TRUE;
+    wind->resize.configure = SDL_TRUE;
+    Wayland_HandlePendingResize(window);
+    wind->shell_surface.libdecor.initial_configure_seen = SDL_TRUE;
+
+    window->w = wind->resize.width;
+    window->h = wind->resize.height;
+
+    /* window state */
+    if (!libdecor_configuration_get_window_state(configuration, &window_state)) {
+        window_state = LIBDECOR_WINDOW_STATE_NONE;
+    }
+
+   /* Always send maximized/restored/focus events; if the event is redundant it will
+     * automatically be discarded (see src/events/SDL_windowevents.c).
+     *
+     * No, we do not get minimize events from libdecor.
+     */
+    if (!(window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN)) {
+        window->flags &= ~SDL_WINDOW_FULLSCREEN_DESKTOP;
+        SDL_SendWindowEvent(window,
+                            (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) ?
+                                SDL_WINDOWEVENT_MAXIMIZED :
+                                SDL_WINDOWEVENT_RESTORED,
+                            0, 0);
+    } else {
+        window->flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
+    }
+    SDL_SendWindowEvent(window,
+                        (window_state & LIBDECOR_WINDOW_STATE_ACTIVE) ?
+                            SDL_WINDOWEVENT_FOCUS_GAINED :
+                            SDL_WINDOWEVENT_FOCUS_LOST,
+                        0, 0);
+
+    /* commit frame state */
+    state = libdecor_state_new(width, height);
+    libdecor_frame_commit(frame, state, configuration);
+    libdecor_state_free(state);
+
+    /* Update the resize capability. Since this will change the capabilities and
+     * commit a new frame state with the last known content dimension, this has
+     * to be called after the new state has been commited and the new content
+     * dimensions were updated. */
+    Wayland_SetWindowResizable(SDL_GetVideoDevice(), window,
+                               window->flags & SDL_WINDOW_RESIZABLE);
+}
 
+static void
+decoration_frame_close(struct libdecor_frame *frame, void *user_data)
+{
+    SDL_SendWindowEvent(((SDL_WindowData *)user_data)->sdlwindow, SDL_WINDOWEVENT_CLOSE, 0, 0);
+}
+
+static void
+decoration_frame_commit(struct libdecor_frame *frame, void *user_data)
+{
+    SDL_WindowData *wind = user_data;
+
+    if (wind->sdlwindow->flags & SDL_WINDOW_OPENGL) {
+        SDL_GL_SwapWindow(wind->sdlwindow);
+    }
+    else if (wind->sdlwindow->flags & SDL_WINDOW_VULKAN) {
+        wl_surface_commit(wind->surface);
+    }
+    else {
+        wl_surface_commit(wind->surface);
+    }
```

Comment:
Minor style fix:
```suggestion
    if (wind->sdlwindow->flags & SDL_WINDOW_OPENGL) {
        SDL_GL_SwapWindow(wind->sdlwindow);
    } else {
        wl_surface_commit(wind->surface);
    }
```

Reply:
Hm, this removed the case for `SDL_WINDOW_VULKAN`? This was required for Capsized I think. Not sure if the original issue now disappeared through a different change.

Reply:
Ah, never mind. You just combined the two cases. I left it this way in case there is a "proper way" in Vulkan to force a surface update.